### PR TITLE
[Paddle Tensor No.23] 新增`TensorA[mask] = TensorB`

### PR DESCRIPTION
**Pull request by LittleHeroZZZX**
### PR Category

User Experience

### PR Types

New features

### Description

Paddle Tensor 规范化：修改`Tensor.__setitem__`的 C++ 实现，实现当 index 是布尔类型，且右侧的输入是 Tensor 时的计算逻辑。

### CODE REVIEW

File: paddle/fluid/pybind/eager_method.cc [hunk 1]
```diff
@@ -1864,21 +1864,8 @@ static PyObject* tensor__setitem_dygraph(TensorObject* self,
...
-            where__ad_func(logical_not_ad_func(transed_index[0]),
-                           transed_sub_tensor,
-                           value_tensor);
-      } else {
-        transed_sub_tensor =
-            index_put__ad_func(transed_sub_tensor, transed_index, value_tensor);
-      }
-
+      transed_sub_tensor =
+          index_put__ad_func(transed_sub_tensor, transed_index, value_tensor);
```
Comment (HydrogenSulfate) [inline paddle/fluid/pybind/eager_method.cc]: 跟切片API负责人讨论了一下，结论是：
1. 原来的where优化是有BUG的，v在不是单个元素的张量的前提下，极大概率不能跟x[mask]进行where计算
2. 原先这样写是为了优化 `x[mask] = 0-D/1-D Tensor`，即“将单个值赋值给x中mask为True的位置上”，原先的写法就可以把v广播到x的形状上，然后以`out = where(mask, x, broadcasted_v)`作为结果，这样速度会比`index_put`快一些。

所以是否可以把此处的代码保留，但是把逻辑（判断条件）改一下，让仅含有单个元素的v可以执行

Comment (LittleHeroZZZX) [reply paddle/fluid/pybind/eager_method.cc]: 好嘞

**Comment by HydrogenSulfate** [comment]
@LittleHeroZZZX 我想了下，单测里是否可以考虑v本身可以广播的情况？如下所示

``` py
import torch

x = torch.zeros(2, 2, 3)
mask = torch.tensor(
    [
        [True, False],
        [False, True],
    ],
)

y = torch.arange(1, 4, dtype=torch.float32)

print(x[0, 0])
print(x[0, 1])
print(x[1, 0])
print(x[1, 1])
print(mask)
x[mask] = y
print(x[0, 0])
print(x[0, 1])
print(x[1, 0])
print(x[1, 1])
```

可以测试下你这个代码是否能处理`v([3])`被广播到`x[mask]形状([2,3])`这种case，如果可以的话，还麻烦在单测里加一个这种广播的测试（可以叫`test_boolean_mask_tensor_broadcast_v`）。然后基本就没啥问题了，如果本地测试OK的话，可以新提交一个PR单独加个单测，这个PR我可以先合入了，不然重新跑CI比较麻烦

**Comment by LittleHeroZZZX** [comment]
> @LittleHeroZZZX 我想了下，单测里是否可以考虑v本身可以广播的情况？如下所示
> 
> ```python
> import torch
> 
> x = torch.zeros(2, 2, 3)
> mask = torch.tensor(
>     [
>         [True, False],
>         [False, True],
>     ],
> )
> 
> y = torch.arange(1, 4, dtype=torch.float32)
> 
> print(x[0, 0])
> print(x[0, 1])
> print(x[1, 0])
> print(x[1, 1])
> print(mask)
> x[mask] = y
> print(x[0, 0])
> print(x[0, 1])
> print(x[1, 0])
> print(x[1, 1])
> ```
> 
> 可以测试下你这个代码是否能处理`v([3])`被广播到`x[mask]形状([2,3])`这种case，如果可以的话，还麻烦在单测里加一个这种广播的测试（可以叫`test_boolean_mask_tensor_broadcast_v`）。然后基本就没啥问题了，如果本地测试OK的话，可以新提交一个PR单独加个单测，这个PR我可以先合入了，不然重新跑CI比较麻烦

这个case测试了没问题，与 PyTorch行为一样。

**Comment by HydrogenSulfate** [comment]
> > @LittleHeroZZZX 我想了下，单测里是否可以考虑v本身可以广播的情况？如下所示
> > ```python
> > import torch
> > 
> > x = torch.zeros(2, 2, 3)
> > mask = torch.tensor(
> >     [
> >         [True, False],
> >         [False, True],
> >     ],
> > )
> > 
> > y = torch.arange(1, 4, dtype=torch.float32)
> > 
> > print(x[0, 0])
> > print(x[0, 1])
> > print(x[1, 0])
> > print(x[1, 1])
> > print(mask)
> > x[mask] = y
> > print(x[0, 0])
> > print(x[0, 1])
> > print(x[1, 0])
> > print(x[1, 1])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 可以测试下你这个代码是否能处理`v([3])`被广播到`x[mask]形状([2,3])`这种case，如果可以的话，还麻烦在单测里加一个这种广播的测试（可以叫`test_boolean_mask_tensor_broadcast_v`）。然后基本就没啥问题了，如果本地测试OK的话，可以新提交一个PR单独加个单测，这个PR我可以先合入了，不然重新跑CI比较麻烦
> 
> 这个case测试了没问题，与 PyTorch行为一样。

好的，那我先合入了，后面麻烦再提交一个PR，补充一个广播的单测

**Review by HydrogenSulfate — APPROVED** [review]
LGTM

**Comment by luotao1** [comment]
hi, @LittleHeroZZZX 
- 非常感谢你对飞桨的贡献，我们正在运营一个PFCC组织，会通过定期分享技术知识与发布开发者主导任务的形式持续为飞桨做贡献，详情可见 https://github.com/luotao1 主页说明。
- 如果你对PFCC有兴趣，请发送邮件至 [ext_paddle_oss@baidu.com](mailto:ext_paddle_oss@baidu.com)，我们会邀请你加入~
